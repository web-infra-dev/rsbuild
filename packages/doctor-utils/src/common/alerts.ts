import { Rule, SDK } from '@rsbuild/doctor-types';
import { relative } from 'path';

export function getPackageRelationAlertDetails(
  modules: SDK.ModuleGraphData['modules'],
  dependencies: SDK.ModuleGraphData['dependencies'],
  root: string,
  packageDependencies: Rule.DependencyWithPackageData[],
  moduleCodeMap: SDK.ModuleCodeData,
): SDK.ServerAPI.InferResponseType<SDK.ServerAPI.API.GetPackageRelationAlertDetails> {
  return packageDependencies
    .slice()
    .reverse()
    .map((dep) => {
      const dependency = dependencies.find(
        (item) => item.id === dep.dependencyId,
      );

      if (!dependency) {
        return null!;
      }

      const module = modules.find((item) => item.id === dependency.module);

      if (!module) {
        return null!;
      }

      return {
        group: dep.group,
        module,
        dependency,
        relativePath: relative(root, module.path),
        moduleCode: moduleCodeMap?.[module.id],
      };
    })
    .filter(Boolean);
}
