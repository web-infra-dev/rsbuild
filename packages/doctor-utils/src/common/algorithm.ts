import { deflateSync, inflateSync } from 'zlib';
import { Buffer } from 'buffer';

export function mergeIntervals(intervals: [number, number][]) {
  // Sort from small to large
  intervals.sort((a, b) => a[0] - b[0]);
  // The previous interval, the next interval, store the result
  let previous;
  let current;
  const result = [];
  for (let i = 0; i < intervals.length; i++) {
    current = intervals[i];
    // If the first interval or the current interval does not overlap with the previous interval, add the current interval to the result
    if (!previous || current[0] > previous[1]) {
      // Assign the current interval to the previous interval
      previous = current;
      result.push(current);
    } else {
      // Otherwise, the two intervals overlap
      // Update the end time of the previous interval
      previous[1] = Math.max(previous[1], current[1]);
    }
  }
  return result;
}

export function compressText(input: string): string {
  return deflateSync(input).toString('base64');
}

export function decompressText(input: string): string {
  return inflateSync(Buffer.from(input, 'base64')).toString();
}

export function random(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}
