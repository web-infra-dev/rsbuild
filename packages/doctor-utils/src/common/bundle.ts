import { Client } from '@rsbuild/doctor-types';

const sep = ',';

export function getBundleDiffPageQueryString(files: string[]) {
  let qs = encodeURIComponent(files.join(sep));

  if (qs) {
    qs = `?${Client.DoctorClientUrlQuery.BundleDiffFiles}=${qs}`;
  }

  return qs;
}

export function getBundleDiffPageUrl(files: string[]) {
  let qs = getBundleDiffPageQueryString(files);

  if (process.env.NODE_ENV === 'development') {
    if (typeof location !== 'undefined') {
      const { search = '', origin } = location;
      if (search) {
        qs += `&${search.slice(1)}`;
      }
      return `${origin}${qs}#${Client.DoctorClientRoutes.BundleDiff}`;
    }
  }

  return `${qs}#${Client.DoctorClientRoutes.BundleDiff}`; // TODO: client host check
}

export function parseFilesFromBundlePageUrlQuery(queryValue: string) {
  return decodeURIComponent(queryValue).split(sep);
}
