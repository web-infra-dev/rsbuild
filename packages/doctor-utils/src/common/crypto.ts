const sep = '!';

export function encode(str: string) {
  let res = `${str.charCodeAt(0)}`;

  for (let i = 1; i < str.length; i++) {
    res += `${sep}${str.charCodeAt(i)}`;
  }

  return res;
}

export function decode(str: string) {
  let res = '';
  let tmp = '';

  for (let i = 0; i < str.length; i++) {
    if (str[i] === sep) {
      res += String.fromCharCode(+tmp);
      tmp = '';
    } else {
      // 'number'
      tmp += str[i];
    }
  }

  // end of str
  if (tmp) {
    res += String.fromCharCode(+tmp);
  }

  return res;
}
