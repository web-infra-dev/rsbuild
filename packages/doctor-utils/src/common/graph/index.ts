export * from './assets';
export * from './chunk';
export * from './modules';
export * from './dependency';
export * from './entrypoints';
