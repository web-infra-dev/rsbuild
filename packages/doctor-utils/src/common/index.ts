export * as Summary from './summary';
export * as Crypto from './crypto';
export * as Manifest from './manifest';
export * as Loader from './loader';
export * as Time from './time';
export * as Algorithm from './algorithm';
export * as Resolver from './resolver';
export * as Graph from './graph';
export * as Bundle from './bundle';
export * as Url from './url';
export * as Plugin from './plugin';
export * as Data from './data';
export * as Alerts from './alerts';
export * as Rspack from './rspack';
export * as Package from './package';
