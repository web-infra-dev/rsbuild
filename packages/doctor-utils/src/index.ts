export * from './build';
export * from './common';
