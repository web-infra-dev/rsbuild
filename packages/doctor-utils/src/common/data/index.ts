import { Rule, SDK, Manifest } from '@rsbuild/doctor-types';

import * as Loader from '../loader';
import * as Resolver from '../resolver';
import * as Plugin from '../plugin';
import * as Graph from '../graph';
import * as Alerts from '../alerts';

/**
 * this class will run at both brower and node environment.
 */
export class APIDataLoader {
  constructor(protected loader: Manifest.ManifestDataLoader) {
    this.loadAPI = this.loadAPI.bind(this);
  }

  public log(...args: unknown[]) {
    console.log(`[${this.constructor.name}]`, ...args);
  }

  public loadAPI<
    T extends SDK.ServerAPI.API,
    B extends
      SDK.ServerAPI.InferRequestBodyType<T> = SDK.ServerAPI.InferRequestBodyType<T>,
    R extends
      SDK.ServerAPI.InferResponseType<T> = SDK.ServerAPI.InferResponseType<T>,
  >(
    ...args: B extends void ? [api: T] : [api: T, body: B]
  ): Promise<SDK.ServerAPI.InferResponseType<T>> {
    const [api, body] = args;
    switch (api) {
      case SDK.ServerAPI.API.LoadDataByKey:
        return this.loader.loadData(
          (
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.LoadDataByKey>
          ).key,
        ) as Promise<R>;

      /** Project API */
      case SDK.ServerAPI.API.GetProjectInfo:
        return Promise.all([
          this.loader.loadData('root'),
          this.loader.loadData('pid'),
          this.loader.loadData('hash'),
          this.loader.loadData('summary'),
          this.loader.loadData('configs'),
          this.loader.loadData('envinfo'),
          this.loader.loadData('errors'),
        ]).then(
          ([root, pid, hash, summary, configs, envinfo, errors]) =>
            ({ root, pid, hash, summary, configs, envinfo, errors }) as R,
        );
      case SDK.ServerAPI.API.GetClientRoutes:
        return this.loader.loadManifest().then((res) => {
          const { enableRoutes = [] } = res.client || {};
          return enableRoutes as R;
        });

      /** Loader API */
      case SDK.ServerAPI.API.GetLoaderNames:
        return this.loader.loadData('loader').then((res) => {
          return Loader.getLoaderNames(res || []) as R;
        });

      case SDK.ServerAPI.API.GetLoaderChartData:
        return this.loader.loadData('loader').then((res) => {
          return Loader.getLoaderChartData(res || []) as R;
        });

      case SDK.ServerAPI.API.GetLoaderFileTree:
        return this.loader.loadData('loader').then((res) => {
          return Loader.getLoaderFileTree(res || []) as R;
        });

      case SDK.ServerAPI.API.GetLoaderFileDetails:
        return this.loader.loadData('loader').then((res) => {
          return Loader.getLoaderFileDetails(
            (
              body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetLoaderFileDetails>
            ).path,
            res || [],
          ) as R;
        });

      case SDK.ServerAPI.API.GetLoaderFolderStatistics:
        return this.loader.loadData('loader').then((res) => {
          return Loader.getLoaderFolderStatistics(
            (
              body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetLoaderFolderStatistics>
            ).folder,
            res || [],
          ) as R;
        });

      case SDK.ServerAPI.API.GetLoaderFileFirstInput:
        return this.loader.loadData('loader').then((res) => {
          return Loader.getLoaderFileFirstInput(
            (
              body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetLoaderFileFirstInput>
            ).file,
            res || [],
          ) as R;
        });

      case SDK.ServerAPI.API.GetLoaderFileInputAndOutput:
        return this.loader.loadData('loader').then((res) => {
          return Loader.getLoaderFileFirstInput(
            (
              body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetLoaderFileFirstInput>
            ).file,
            res || [],
          ) as R;
        });

      /** Resolver API */
      case SDK.ServerAPI.API.GetResolverFileTree:
        return this.loader.loadData('resolver').then((res) => {
          return Resolver.getResolverFileTree(res || []) as R;
        });
      case SDK.ServerAPI.API.GetResolverFileDetails:
        return Promise.all([
          this.loader.loadData('resolver'),
          this.loader.loadData('moduleGraph.modules'),
          this.loader.loadData('moduleCodeMap'),
        ]).then((res) => {
          const resolverData = res[0];
          const modules = res[1];
          const moduleCodeMap = res[2];

          return Resolver.getResolverFileDetails(
            (
              body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetResolverFileDetails>
            ).filepath,
            resolverData || [],
            modules || [],
            moduleCodeMap || {},
          ) as R;
        });

      /** Plugin API */
      case SDK.ServerAPI.API.GetPluginSummary:
        return this.loader.loadData('plugin').then((res) => {
          return Plugin.getPluginSummary(res || {}) as R;
        });
      case SDK.ServerAPI.API.GetPluginData:
        return this.loader.loadData('plugin').then((res) => {
          const { hooks, tapNames } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetPluginData>;
          return Plugin.getPluginData(res || {}, hooks, tapNames) as R;
        });

      /** Graph API */
      case SDK.ServerAPI.API.GetAssetsSummary:
        return this.loader.loadData('chunkGraph').then((res) => {
          const { withFileContent = true } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetAssetsSummary>;
          const { assets = [], chunks = [] } = res || {};
          return Graph.getAssetsSummary(assets, chunks, {
            withFileContent,
          }) as R;
        });
      case SDK.ServerAPI.API.GetAssetDetails:
        return Promise.all([
          this.loader.loadData('chunkGraph'),
          this.loader.loadData('moduleGraph'),
        ]).then((res) => {
          const { assetPath } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetAssetDetails>;
          const { assets = [], chunks = [] } = res[0] || {};
          const { modules = [] } = res[1] || {};
          return Graph.getAssetDetails(assetPath, assets, chunks, modules) as R;
        });
      case SDK.ServerAPI.API.GetChunksByModuleId:
        return Promise.all([
          this.loader.loadData('chunkGraph'),
          this.loader.loadData('moduleGraph'),
        ]).then((res) => {
          const { moduleId } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetChunksByModuleId>;
          const { chunks = [] } = res[0] || {};
          const { modules = [] } = res[1] || {};
          return Graph.getChunksByModuleId(moduleId, modules, chunks) as R;
        });
      case SDK.ServerAPI.API.GetModuleDetails:
        return Promise.all([
          this.loader.loadData('chunkGraph'),
          this.loader.loadData('moduleGraph'),
        ]).then((res) => {
          const { moduleId } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetModuleDetails>;
          const { chunks = [] } = res[0] || {};
          const { modules = [], dependencies = [] } = res[1] || {};
          return Graph.getModuleDetails(moduleId, modules, dependencies) as R;
        });
      case SDK.ServerAPI.API.GetModulesByModuleIds:
        return this.loader.loadData('moduleGraph').then((res) => {
          const { moduleIds } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetModulesByModuleIds>;
          const { modules = [] } = res || {};
          return Graph.getModuleIdsByModulesIds(moduleIds, modules) as R;
        });

      case SDK.ServerAPI.API.GetEntryPoints:
        return Promise.all([this.loader.loadData('chunkGraph')]).then((res) => {
          const [chunkGraph] = res;
          const { entrypoints = [] } = chunkGraph || {};

          return Graph.getEntryPoints(entrypoints) as R;
        });

      case SDK.ServerAPI.API.GetModuleCodeByModuleId:
        return this.loader.loadData('moduleCodeMap').then((moduleCodeMap) => {
          const { moduleId } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetModuleCodeByModuleId>;
          if (moduleCodeMap) {
            return moduleCodeMap[moduleId] as R;
          }
          return { source: '', transformed: '', parsedSource: '' } as R;
        });

      case SDK.ServerAPI.API.GetModuleCodeByModuleIds:
        return this.loader.loadData('moduleCodeMap').then((moduleCodeMap) => {
          const { moduleIds } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetModuleCodeByModuleIds>;
          const _moduleCodeData = {} as R;
          if (moduleCodeMap) {
            moduleIds.forEach((id: number) => {
              _moduleCodeData[id] = moduleCodeMap[id] as R;
            });
            return _moduleCodeData as R;
          }
          return [] as R;
        });

      case SDK.ServerAPI.API.GetAllModuleGraph:
        return this.loader.loadData('moduleGraph').then((moduleGraph) => {
          return moduleGraph?.modules as R;
        });

      case SDK.ServerAPI.API.GetAllChunkGraph:
        return this.loader.loadData('chunkGraph').then((chunkGraph) => {
          return chunkGraph?.chunks as R;
        });

      case SDK.ServerAPI.API.GetPackageRelationAlertDetails:
        return Promise.all([
          this.loader.loadData('moduleGraph'),
          this.loader.loadData('errors'),
          this.loader.loadData('root'),
          this.loader.loadData('moduleCodeMap'),
        ]).then((res) => {
          const { id, target } =
            body as SDK.ServerAPI.InferRequestBodyType<SDK.ServerAPI.API.GetPackageRelationAlertDetails>;
          const [moduleGraph, errors = [], root = '', moduleCodeMap] = res;
          const { modules = [], dependencies = [] } = moduleGraph || {};
          const { packages = [] } = (errors.find((e) => e.id === id) ||
            {}) as Rule.PackageRelationDiffRuleStoreData;

          const { dependencies: pkgDependencies = [] } =
            packages.find(
              (e) =>
                e.target.name === target.name &&
                e.target.root === target.root &&
                e.target.version === target.version,
            ) || {};

          return Alerts.getPackageRelationAlertDetails(
            modules,
            dependencies,
            root,
            pkgDependencies,
            moduleCodeMap || {},
          ) as R;
        });

      case SDK.ServerAPI.API.GetOverlayAlerts:
        return this.loader.loadData('errors').then((res) => {
          return (res || []).filter(
            (e) => e.code === Rule.RuleMessageCodeEnumerated.Overlay,
          ) as R;
        });

      /** Bundle Diff API */
      case SDK.ServerAPI.API.BundleDiffManifest:
        return this.loader.loadManifest() as Promise<R>;

      case SDK.ServerAPI.API.GetBundleDiffSummary:
        return Promise.all([
          this.loader.loadManifest(),
          this.loader.loadData('root'),
          this.loader.loadData('hash'),
          this.loader.loadData('errors'),
          this.loader.loadData('chunkGraph'),
          this.loader.loadData('moduleGraph'),
          this.loader.loadData('moduleCodeMap'),
          this.loader.loadData('packageGraph'),
          this.loader.loadData('configs'),
        ]).then(
          ([
            manifest,
            root = '',
            hash = '',
            errors = {},
            chunkGraph = {},
            moduleGraph = {},
            moduleCodeMap = {},
            packageGraph = {},
            configs = [],
          ]) => {
            let outputFilename = '';

            // outputFile is user's repo build config: output.filename. Detial: https://webpack.docschina.org/configuration/output/#outputfilename.
            if (typeof configs[0].config?.output?.chunkFilename === 'string') {
              outputFilename = configs[0].config.output.chunkFilename;
            }

            return {
              root,
              hash,
              errors,
              chunkGraph,
              moduleGraph,
              packageGraph,
              outputFilename,
              moduleCodeMap,
            } as R;
          },
        );

      case SDK.ServerAPI.API.GetTileReportHtml:
        return this.loader.loadData('otherReports').then((otherReports) => {
          return otherReports?.tileReportHtml as R;
        });

      default:
        throw new Error(`API not implement: "${api}"`);
    }
  }
}
