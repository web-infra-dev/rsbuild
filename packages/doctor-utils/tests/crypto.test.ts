import { describe, it, expect } from 'vitest';
import { Crypto } from '../src/common';

describe('test src/crypto.ts', () => {
  describe('encode & decode', () => {
    const strings = [
      '00123sdfdkfkskfk',
      '~!ï½œ.[]{}&â€¦â€¦%ï¼ˆï¼‰()`Â·sdkf121',
      '~sdï½žd8å“ˆå“ˆå“ˆðŸ˜Š',
      'ðŸ”šðŸ”œðŸŒ¹ðŸ‘Œ',
    ];

    it('encode()', () => {
      expect(
        strings.map((e) => `"${e}" -> "${Crypto.encode(e)}"`).join('\n'),
      ).toMatchSnapshot();
    });

    it('decode()', () => {
      const res = strings
        .map((e) => Crypto.encode(e))
        .map((e) => Crypto.decode(e));
      expect(res).toStrictEqual(strings);
    });
  });
});
