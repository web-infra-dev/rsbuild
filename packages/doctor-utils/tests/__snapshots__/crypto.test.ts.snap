// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test src/crypto.ts > encode & decode > encode() 1`] = `
""00123sdfdkfkskfk" -> "48!48!49!50!51!115!100!102!100!107!102!107!115!107!102!107"
"~!ï½œ.[]{}&â€¦â€¦%ï¼ˆï¼‰()\`Â·sdkf121" -> "126!33!65372!46!91!93!123!125!38!8230!8230!37!65288!65289!40!41!96!183!115!100!107!102!49!50!49"
"~sdï½žd8å“ˆå“ˆå“ˆðŸ˜Š" -> "126!115!100!65374!100!56!21704!21704!21704!55357!56842"
"ðŸ”šðŸ”œðŸŒ¹ðŸ‘Œ" -> "55357!56602!55357!56604!55356!57145!55357!56396""
`;
